pilot/pkg/bootstrap/certcontroller.go

```diff
@@ -126,10 +93,94 @@ func (s *Server) initCertController(args *PilotArgs) error {
 	return nil
 }
 
-func pilotDNSCertDir() (string, error) {
-	userHomeDir, err := os.UserHomeDir()
+// initDNSCerts will create the certificates to be used by Istiod GRPC server and webhooks, signed by K8S server.
+// If the certificate creation fails - for example no support in K8S - returns an error.
+// Will use the mesh.yaml DiscoveryAddress to find the default expected address of the control plane,
+// with an environment variable allowing override.
+//
+// TODO: If the discovery address in mesh.yaml is set to port 15012 (XDS-with-DNS-certs) and the name
+// matches the k8s namespace, failure to start DNS server is a fatal error.
+func (s *Server) initDNSCerts() error {
+	if _, err := os.Stat(DNSCertDir + "/key.pem"); err == nil {
+		// Existing certificate mounted by user. Skip self-signed certificate generation.
+		// Use this with an existing CA - the expectation is that the cert will match the
+		// DNS name in DiscoveryAddress.
+		return nil
+	}
+
+	// discAddr configured in mesh config - this is what we'll inject into pods.
+	discAddr := s.mesh.DefaultConfig.DiscoveryAddress
+
+	// Explicit override - discovery address may point to a remote cluster.
+	// Istiod will be a local service doing only injection
+	if IstiodAddress.Get() != "" {
+		discAddr = IstiodAddress.Get()
+	}
+	host, port, err := net.SplitHostPort(discAddr)
 	if err != nil {
-		return "", fmt.Errorf("could not find local user folder: %v", err)
+		log.Fatala("Invalid discovery address", discAddr, err)
+	}
+
+	hostParts := strings.Split(host, ".")
+
+	ns := "." + PodNamespaceVar.Get()
+
+	envName := hostParts[0]
+
+	// Names in the Istiod cert - support the old service names as well.
+	// The first is the recommended one, also used by Apiserver for webhooks.
+	names := []string{
+		envName + ns + ".svc",
+		envName + ns}
+
+	// Special case - this is the default control plane ( istio-pilot ), take over the role of
+	// CA signer and galley.
+	if envName == "istio-pilot" || envName == "istiod" {
+		if envName == "istio-pilot" {
+			names = append(names, "istiod"+ns)
+		} else {
+			names = append(names, "istio-pilot"+ns)
+		}
+		names = append(names, "istio-galley"+ns)
+		names = append(names, "istio-ca"+ns)
+	}
+
+	log.Infoa("Generating K8S-signed cert for ", names)
```

Comment:
Yes, they match what K8S uses for the injector and what we use internally in mesh config for discoveryAddress.
They must match - otherwise certs don't work.

For FQDN names - you'll need to get certs from a real CA.

Reply:
I mean can we access pilot by  `istio-pilot.namespace.svc.cluster.local`?

Reply:
No, this will not work, and we didn't use this before. 

We could add aliases to the cert - but so far the primary name we need to support is the .svc name which is auto-generated by k8s for injection.

Reply:
And instead of supporting 5 different options - in this case it makes sense to standardize on one.